### ISSUE TITLE
Warn for variables not being closed over by roles

### ISSUE THREAD

**Issue by vrurg** [rakudo/rakudo#2882]
## The Problem

When a class or a role created by a routine using closure the new type object might get its closure from a wrong thread if the routine is run in parallel.

## Steps to Reproduce

The following code demonstrates the problem:

```
sub via-role ($obj, $name) {
    $obj but my role :: { method test-name {$name} }
}

sub via-class ($obj, $name) {
    my class Bar {
        method test-name {$name}
    }.new
}

class Foo {
    method test-name { "default" }
}

my @p;
for 1..100 -> $i {
    @p.push: start {
        $*ERR.print: "$i\r";
        my $obj = Foo.new;
        my $name = "name$i";
        my $robj = via-role($obj, $name);
        my $cobj = via-class($obj, $name);
        say $name, " => ", $robj.test-name, " // ", $robj.^name unless $name eq $robj.test-name;
        say $name, " => ", $cobj.test-name, " // ", $cobj.^name unless $name eq $cobj.test-name;
        sleep .01.rand;
    }
}

await @p;

```

Output of the script might look like:

>```
>name45 => name46 // Foo+{<anon|1>}
>name45 => name46 // Bar
>name50 => name51 // Foo+{<anon|1>}
>name50 => name51 // Bar
>name54 => name55 // Foo+{<anon|1>}
>name54 => name55 // Bar
>name79 => name78 // Foo+{<anon|1>}
>name79 => name78 // Bar
>name89 => name90 // Foo+{<anon|1>}
>name89 => name90 // Bar
>```

An interesting observation: they always come in pairs. I.e. if `via-role` fires then `via-class` will fire two with the exact same values.

Routine closures are not suffering from this problem.

Sometimes the script might finish with no output. With 1000 repetitions never seen it passing without printing a couple of lines.

## Environment
* Operating system: macOS, Linux
* Compiler version (`perl6 -v`): This is Rakudo version 2019.03.1-348-gf2ccd6fa8 built on MoarVM version 2019.03-113-g5047aebb4

Same effect observed for 2019.03, 2018.11.

**Comment by vrurg** [comment]
As @ugexe reminded me, classes/roles don't close over variables. But at least a warning would be nice to have.

**Comment by lizmat** [comment]
I'm pretty sure that would be a very hard thing to do.  @jnthn: do you agree?

**Comment by vrurg** [comment]
Very hard to add a warning? When the grammar encounters a symbol it looks it up the in the enclosing contexts, if I'm not mistaken. Is it so hard to check that the symbol is a lexical from another package?

**Comment by jnthn** [comment]
I'm pretty sure there's another issue asking pretty much the same question somewhere. Anyway, I think a warning is viable; only question is if it makes unacceptable noise for any existing code, so we view it as wanting a language guard. But I'd say let's try without, and see what blin tells us.

**Comment by jnthn** [comment]
Ah, the older issue is https://github.com/rakudo/rakudo/issues/2690 and it has some useful comments on it.

**Comment by vrurg** [comment]
BTW, as to the language version – I was about to ask you to review #2852 in the module loading aspect – I'm not sure if CORE.<letter>.setting is being loaded the right way. spectest is passing for it except for some 6.d.PREVIEW tests.

**Comment by lizmat** [comment]
Closing in favour of #2690
